extensions/interactions/GraphInput/directives/GraphInput.js

```diff
@@ -255,49 +283,70 @@ oppia.directive('graphViz', [
-              $scope.graph.vertices[$scope.state.currentlyDraggedVertex].y =
-                $scope.state.vertexDragStartY + (
-                  $scope.state.mouseY - $scope.state.mouseDragStartY);
+            if ($scope.interactionIsActive) {
+              $scope.state.mouseX = event.pageX - vizContainer.offset().left;
+              $scope.state.mouseY = event.pageY - vizContainer.offset().top;
+              // We use vertexDragStartX/Y and mouseDragStartX/Y to make
+              // mouse-dragging by label more natural, by moving the vertex
+              // according to the difference from the original position.
+              // Otherwise, mouse-dragging by label will make the vertex
+              // awkwardly jump to the mouse.
+              if ($scope.state.currentlyDraggedVertex !== null &&
+                  ($scope.state.mouseX > GRAPH_INPUT_LEFT_MARGIN)) {
+                $scope.graph.vertices[$scope.state.currentlyDraggedVertex].x =
+                  $scope.state.vertexDragStartX + (
+                    $scope.state.mouseX - $scope.state.mouseDragStartX);
+                $scope.graph.vertices[$scope.state.currentlyDraggedVertex].y =
+                  $scope.state.vertexDragStartY + (
+                    $scope.state.mouseY - $scope.state.mouseDragStartY);
+              }
             }
           };
 
           $scope.onClickGraphSVG = function() {
-            if ($scope.state.currentMode === _MODES.ADD_VERTEX &&
-                $scope.canAddVertex) {
-              $scope.graph.vertices.push({
-                x: $scope.state.mouseX,
-                y: $scope.state.mouseY,
-                label: ''
-              });
-              setMode(_MODES.MOVE);
-            }
-            if ($scope.state.hoveredVertex === null) {
-              $scope.state.selectedVertex = null;
-            }
-            if ($scope.state.hoveredEdge === null) {
-              $scope.state.selectedEdge = null;
+            if ($scope.interactionIsActive) {
+              if ($scope.state.currentMode === _MODES.ADD_VERTEX &&
+                  $scope.canAddVertex) {
+                $scope.graph.vertices.push({
+                  x: $scope.state.mouseX,
+                  y: $scope.state.mouseY,
+                  label: ''
+                });
+                setMode(_MODES.MOVE);
+              }
+              if ($scope.state.hoveredVertex === null) {
+                $scope.state.selectedVertex = null;
+              }
+              if ($scope.state.hoveredEdge === null) {
+                $scope.state.selectedEdge = null;
+              }
             }
           };
 
           $scope.init = function() {
             initButtons();
-            $scope.state.currentMode = $scope.buttons[0].mode;
+            if ($scope.buttons.length > 0) {
+              $scope.state.currentMode = $scope.buttons[0].mode;
+            } else {
+              $scope.interactionIsActive = false;
```

Comment:
I don't understand why we're setting $scope.interactionIsActive based on the length of the buttons array. Could you elaborate?

It seems more like the first should cause the second, rather than the other way round.

Reply:
In the init at the beginning odf the file, I don't set the buttons at all if he is revisiting the interaction and so based on that I am setting interactionIsActive in this directive.

Reply:
That's still conceptually clunky, though. See my comment in CodeRepl.js for another possible way to handle this.

Reply:
Done, I added another attribute called canEditGraph and using its value I set interactionIsActive in the directive.